language-support/ts/codegen/src/TsCodeGenMain.hs

```diff
@@ -55,14 +50,46 @@ optionsParserInfo = info (optionsParser <**> helper)
 main :: IO ()
 main = do
     opts@Options{..} <- execParser optionsParserInfo
-    dar <- B.readFile optInputDar
-    let archive = Zip.toArchive $ BSL.fromStrict dar
-    dalfs <- either fail pure $ DAR.readDalfs archive
-    DAR.DalfManifest{packageName, ..} <- either fail pure $ DAR.readDalfManifest archive
-    packageName <- pure $ optMainPackageName <|> packageName
-    forM_ ((DAR.mainDalf dalfs, packageName) : map (, Nothing) (DAR.dalfs dalfs)) $ \(dalf, mbPkgName) -> do
-        (pkgId, pkg) <- either (fail . show)  pure $ Archive.decodeArchive Archive.DecodeAsMain (BSL.toStrict dalf)
-        daml2ts opts pkgId pkg mbPkgName
+    foldM_ (processDar opts) Map.empty optInputDars
+      where
+        -- Generate the ts for a single DAR. 'processed' is a map of
+        -- package ids of processed DALFs (the same package can appear
+        -- in multiple DARS - avoid regenerating them where possible).
+        processDar :: Options -> Map.Map PackageId [String] -> FilePath -> IO (Map.Map PackageId [String])
+        processDar opts processed dar = do
+          dar <- B.readFile dar
+          let archive = Zip.toArchive $ BSL.fromStrict dar
+          dalfs <- either fail pure $ DAR.readDalfs archive
+          DAR.DalfManifest{packageName, ..} <- either fail pure $ DAR.readDalfManifest archive
+          let allDalfsInDar = (DAR.mainDalf dalfs, packageName) : map (, Nothing) (DAR.dalfs dalfs)
+          foldM (processDalf opts) processed allDalfsInDar
+
+        -- Generate the ts for a single DALF. Avoid generating it
+        -- multiple times where possible.
+        processDalf :: Options -> Map.Map PackageId [String] -> (BSL.ByteString, Maybe String) -> IO (Map.Map PackageId [String])
+        processDalf opts pkgs (dalf, mbPkgName) = do
+          (pkgId, pkg) <- either (fail . show)  pure $ Archive.decodeArchive Archive.DecodeAsMain (BSL.toStrict dalf)
+          let mainPkgNames = [x | x <- concat (Map.elems pkgs), not $ null x]
+          gen <- case Map.lookup pkgId pkgs of
+            Nothing -> do
+              maybe (return ()) (\name -> when (name `elem` mainPkgNames) $
```

Comment:
`whenJust` is your friend: https://www.stackage.org/haddock/lts-14.6/extra-1.6.18/Control-Monad-Extra.html#v:whenJust

Reply:
The condition is similar but different depending on the `Just` and `Nothing` cases. Not sure `whenJust` helps here.

Reply:
```suggestion
              whenJust mbPkgName $ \name -> when (name `elem` mainPkgNames) $ fail ...
```